src/od/od.rs

```diff
@@ -74,418 +96,336 @@ pub fn uumain(args: Vec<String>) -> i32 {
-            OdFormat {
-                itembytes: itembytes,
-                writer: fmtspec.writer,
-                offmarg: fmtspec.offmarg,
+    let mut line_bytes = match matches.opt_default("w", "32") {
+        None => 16,
+        Some(s) => {
+            match s.parse::<usize>() {
+                Ok(i) => { i }
+                Err(_) => { 2 }
             }
         }
+    };
+    let min_bytes = formats.iter().fold(1, |max, next| cmp::max(max, next.formatter_item_info.byte_size));
+    if line_bytes % min_bytes != 0 {
+        show_warning!("invalid width {}; using {} instead", line_bytes, min_bytes);
+        line_bytes = min_bytes;
+    }
 
-// TODO: -t fmts
-        let known_formats = hashmap![
-    		"a" => (1, &a_char),
-    		"B" => (2, &oct) ,
-    		"b" => (1, &oct),
-    		"c" => (1, &c_char),
-    		"D" => (4, &dec_u),
-    // TODO: support floats
-    //		"e" => (8, &flo64),
-    //		"F" => (8, &flo64),
-    //		"F" => (4, &flo32),
-    		"H" => (4, &hex),
-    		"X" => (4, &hex) ,
-    		"o" => (2, &oct),
-    		"x" => (2, &hex),
-    		"h" => (2, &hex),
-
-    		"I" => (2, &dec_s),
-    		"L" => (2, &dec_s),
-    		"i" => (2, &dec_s),
-
-    		"O" => (4, &oct),
-    		"s" => (2, &dec_u)
-    	];
-
-        let mut formats = Vec::new();
-
-        for flag in flags.iter() {
-            match known_formats.get(flag) {
-                None => {} // not every option is a format
-                Some(r) => {
-                    let (itembytes, fmtspec) = *r;
-                    formats.push(mkfmt(itembytes, fmtspec))
+    let output_duplicates = matches.opt_present("v");
+
+    let skip_bytes = match matches.opt_default("skip-bytes", "0") {
+        None => 0,
+        Some(s) => {
+            match parse_number_of_bytes(&s) {
+                Ok(i) => { i }
+                Err(_) => {
+                    disp_err!("Invalid argument --skip-bytes={}", s);
+                    return 1;
                 }
             }
         }
-
-        if formats.is_empty() {
-            formats.push(mkfmt(2, &oct)); // 2 byte octal is the default
+    };
+    let read_bytes = match matches.opt_str("read-bytes") {
+        None => None,
+        Some(s) => {
+            match  parse_number_of_bytes(&s) {
+                Ok(i) => { Some(i) }
+                Err(_) => {
+                    disp_err!("Invalid argument --read-bytes={}", s);
+                    return 1;
+                }
+            }
         }
+    };
 
-        odfunc(&input_offset_base, &inputs, &formats[..])
+    odfunc(line_bytes, input_offset_base, byte_order, inputs, &formats[..],
+            output_duplicates, skip_bytes, read_bytes)
 }
 
-const LINEBYTES:usize = 16;
-const WORDBYTES:usize = 2;
+fn odfunc(line_bytes: usize, input_offset_base: Radix, byte_order: ByteOrder,
+        fnames: Vec<InputSource>, formats: &[ParsedFormatterItemInfo], output_duplicates: bool,
+        skip_bytes: usize, read_bytes: Option<usize>) -> i32 {
+
```

Comment:
Because there are now many function inputs in this complex function, it may be good to organize the inputs into a struct, to make it is easy to differentiate within the function between function inputs and derivations based on those.